### ISSUE TITLE
[HOLD for payment 2023-03-31] [$4000] Focus on the field does not shift while editing in the general settings page at the workspace.

### ISSUE THREAD

**Issue by kavimuru** [Expensify/App#15109]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:

1. Open any workspace and go to general settings.
2. Now perform edit on workspace name (Do not save)
3. Notice that the cursor is focused on workspace name.
4. Now change the currency

## Expected Result:
The focus on field should change in android too. I.e when selecting a new field, the cursor should also shift to that new field.

## Actual Result:
Even After editing name field and moving to currency field, the cursor witha green line focus still remains on name field.

## Workaround:
unknown

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [x] Android / native

**Version Number:** 1.2.71-0
**Reproducible in staging?:** y
**Reproducible in production?:** y
**If this was caught during regression testing, add the test name, ID and link from TestRail:**
**Email or phone of affected tester (no customers):**
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Notes/Photos/Videos:** 

https://user-images.githubusercontent.com/43996225/218585723-8350f5f8-6903-4479-9027-6e97bad16fb9.mp4

https://user-images.githubusercontent.com/43996225/218585871-5d88f859-596c-4bb4-8397-221bec517c69.mp4

**Expensify/Expensify Issue URL:**
**Issue reported by:** @ashimsharma10
**Slack conversation:** https://expensify.slack.com/archives/C049HHMV9SM/p1676191167476849

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~017b5fb21ef5da838a</li>
        <li>Upwork Job ID: 1627908747207131136</li>
        <li>Last Price Increase: 2023-03-07</li>
    </ul>
</details>

**Comment by MelvinBot** [comment]
Triggered auto assignment to @anmurali (`Bug`), see https://stackoverflow.com/c/expensify/questions/14418 for more details.

**Comment by MelvinBot** [comment]
### Bug0 Triage Checklist ([Main S/O](https://stackoverflowteams.com/c/expensify/questions/14418))

- [ ] This "bug" occurs on a supported platform (ensure `Platforms` in OP are ✅)
- [ ] This bug is not a duplicate report (check [E/App issues](https://github.com/Expensify/App/issues) and [#expensify-bugs](https://expensify.slack.com/archives/C049HHMV9SM))
    - If it is, comment with a link to the original report, close the issue and add any novel details to the original issue instead
- [ ] This bug is reproducible using the reproduction steps in the OP. [S/O](https://stackoverflowteams.com/c/expensify/questions/14418#:~:text=The%20bug%20is%20reproducible)
    - If the reproduction steps are clear and you're unable to reproduce the bug, check with the reporter and QA first, then close the issue.
    - If the reproduction steps aren't clear and you determine the correct steps, please update the OP.
- [ ] This issue is filled out as thoroughly and clearly as possible
    - Pay special attention to the title, results, platforms where the bug occurs, and if the bug happens on staging/production.
- [ ] I have reviewed and subscribed to the linked Slack conversation to ensure Slack/Github stay in sync

**Comment by MelvinBot** [comment]
@anmurali Whoops! This issue is 2 days overdue. Let's get this updated quick!

**Comment by MelvinBot** [comment]
Job added to Upwork: https://www.upwork.com/jobs/~017b5fb21ef5da838a

**Comment by MelvinBot** [comment]
Current assignee @anmurali is eligible for the External assigner, not assigning anyone new.

**Comment by MelvinBot** [comment]
Triggered auto assignment to Contributor-plus team member for initial proposal review - @thesahindia (`External`)

**Comment by MelvinBot** [comment]
Triggered auto assignment to @tgolen (`External`), see https://stackoverflow.com/c/expensify/questions/7972 for more details.

**Comment by anmurali** [comment]
Reproduced and slapped an external label on this to fix

**Comment by jayeshmangwani** [comment]
@anmurali  is this not expected behavior?
this same behavior happens with all the form pages e.g. VBA > Company information page, focus any text input(Phone number) and then select company type(Picker) after selecting values from Picker, text input is still focused and Picker loses the focus

**Comment by ashimsharma10** [comment]
yes, I guess it occurs in a few more places. but how can such be intended? IMO focus should be on the field where is the user is active.

**Comment by situchan** [comment]
I think we normally respect native platform's behavior. Android native Spinner modal doesn't focus itself at all as default.
Similar issue which was closed: https://github.com/Expensify/App/issues/15230

**Comment by tgolen** [comment]
From my initial read of this situation, I think it does sounds like a bug, especially because we want the following:

1. iOS and Android should behave the same 
2. The UX should be what is expected and not merely what happens due to other "workarounds" that have been put into place

I think the next step here would be to consider if this is a bug that should be fixed higher up in React-Native. Maybe looking at a very basic app (like the react-native testing app) and see how it behaves natively on both platforms would give us a better understanding of what kind of a solution we should be aiming for.

**Comment by devanshujain95** [comment]
@tgolen I totally agree on the points you mentioned regarding same behaviour and expected UX. Although @situchan and other guys know a lot more things for Expensify, but if we have a solution that makes this issue/UX weird behaviour go away, then why are we delaying this?

**Comment by MelvinBot** [comment]
📣 @devanshujain95! 📣

Hey, it seems we don’t have your contributor details yet! You'll only have to do this once, and this is how we'll hire you on Upwork.
Please follow these steps:

1. Get the email address used to login to your Expensify account. If you don't already have an Expensify account, create one [here](https://new.expensify.com/). If you have multiple accounts (e.g. one for testing), please use your main account email.
2. Get the link to your Upwork profile. It's necessary because we only pay via Upwork. You can access it by logging in, and then clicking on your name. It'll look like [this](https://www.upwork.com/freelancers/~01ca8ad8ce36b9eceb). If you don't already have an account, sign up for one [here](https://www.upwork.com/nx/signup/?dest=home).
3. Copy the format below and paste it in a comment on this issue. Replace the placeholder text with your actual details.

![Screen Shot 2022-11-16 at 4 42 54 PM](https://user-images.githubusercontent.com/30609178/202326034-1a292e64-48fd-4fef-9f4f-7feaaeedfd9f.png)

Format:
```
Contributor details
Your Expensify account email: <REPLACE EMAIL HERE>
Upwork Profile Link: <REPLACE LINK HERE>
```

**Comment by devanshujain95** [comment]
Contributor details
Your Expensify account email: devanshujain95@gmail.com
Upwork Profile Link: https://www.upwork.com/freelancers/~01106012e660525565

**Comment by MelvinBot** [comment]
✅ Contributor details stored successfully. Thank you for contributing to Expensify!

**Comment by devanshujain95** [comment]
@thesahindia 
> @devanshujain95, can you explain why it's unable to dismiss the keyboard?

Picker component is unable to handle android's keyboard status. 
Basically for a text input component, to get out of focus or get blur, keyboard should get removed. But our picker component unables to handle or dismiss keyboard, and hence focus remains on TextInput component. This same behaviour can be seen if we clicks on return key present at keyboard, focus gets out and blur event is called. 

> Can you explain how you will do it?

I can explicitly dismiss keyboard everytime Picker component gets in focus. By this way, we can make sure that this weird behaviour will gets fixed at all places without we need to change at any other places.

**Comment by bernhardoj** [comment]
This is the native behavior of Android Picker (Spinner) component. RN picker select uses RN picker under the hood and the picker component on Android use Spinner, an Android dropdown component.

![image](https://user-images.githubusercontent.com/50919443/221363442-a276d673-02be-4041-858b-2d115f24e7b7.png)


https://user-images.githubusercontent.com/50919443/221365360-c65da00d-7653-4f0c-89bd-6036bf7e5ada.mp4

**Comment by MelvinBot** [comment]
@tgolen @anmurali @thesahindia this issue was created 2 weeks ago. Are we close to approving a proposal? If not, what's blocking us from getting this issue assigned? Don't hesitate to create a thread in #expensify-open-source to align faster in real time. Thanks!

**Comment by anmurali** [comment]
@roryabraham - Tim tells him you might have dealt with an upstream fix like this before? Does this make sense to bring to an expert contributor like Margelo? If yes, can you help me figure out next steps?

**Comment by MelvinBot** [comment]
@tgolen @anmurali @thesahindia this issue is now 3 weeks old. There is one more week left before this issue <a href='https://stackoverflowteams.com/c/expensify/questions/15233/15234#15234'>breaks WAQ</a> and will need to go internal. What needs to happen to get a PR in review this week? Please create a thread in #expensify-open-source to discuss. Thanks!

**Comment by MelvinBot** [comment]
@tgolen @anmurali @thesahindia this issue is now 3 weeks old. There is one more week left before this issue <a href='https://stackoverflowteams.com/c/expensify/questions/15233/15234#15234'>breaks WAQ</a> and will need to go internal. What needs to happen to get a PR in review this week? Please create a thread in #expensify-open-source to discuss. Thanks!

**Comment by MelvinBot** [comment]
@tgolen, @anmurali, @thesahindia Uh oh! This issue is overdue by 2 days. Don't forget to update your issues!

**Comment by roryabraham** [comment]
Well, with upstream issues (err... all issues) the first step is always to identify the root source of the problem. In this case, I kind of doubt that the problem originates in React Native itself, since [the Picker component was deprecated](https://reactnative.dev/docs/0.65/picker) several versions ago and has since been [removed](https://github.com/facebook/react-native/blob/24dbb5df0a12eb1bdc1e54375db78a4b4b938871/index.js#L670-L680) from the React Native codebase entirely.

In our app, we use [react-native-picker-select](https://github.com/lawnstarter/react-native-picker-select), which in turn is [dependent upon](https://github.com/lawnstarter/react-native-picker-select/blob/ca6488c2eef5c776a8071400c8b7987712d43397/package.json#L27) the [@react-native-picker/picker](https://github.com/react-native-picker/picker) community package.

So that means the focus problem probably exists in one of those two repos (react-native-picker-select or @react-native-picker/picker).

- The next step is to look for an existing issue in the upstream repo from which the problem originates (I haven't done enough research to know if that's react-native-picker-select or @react-native-picker/picker).
- If an issue describing the problem already exists, but does not have a minimal reproduction example, it is often helpful to create a minimal reproduction, as @tgolen suggested above. If an issue does not already exist, then create a minimal reproduction example and post it along with a detailed description of the problem.
- The next steps depend a lot on:
    - The library
    - The complexity of the problem and related solution
    - How responsive the maintainers of the library are

So I think it would be good to start with just those steps. Oftentimes it seems like a minimal reproduction is half the battle.

**Comment by tgolen** [comment]
I think regardless of it being called a "bug" or not, the goal of RN is to have consistent behavior across platforms, and if that's not happening, then we should make an effort to improve the upstream libraries to have consistent behavior.

**Comment by bernhardoj** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.

When we open a picker on Android, it does not blur the previous text input we focused on.

### What is the root cause of that problem?

So, turns out the picker in our app render a non-editable `TextInput` because we set the `useNativeAndroidPickerStyle` props to `false` which align with the iOS which also render a non-editable `TextInput`. Both non-editable `TextInput` and Android Spinner is not "focusable", nothing different on iOS. But why on iOS, the previous `TextInput` blurs when we open the Picker?

That is because the iOS picker dismisses the Keyboard every time we click it, and it's implemented by `react-native-picker-select` by this PR https://github.com/lawnstarter/react-native-picker-select/pull/125 that solves this issue https://github.com/lawnstarter/react-native-picker-select/issues/123. There is a discussion at the bottom about the Android, but they say it's an upstream issue, but it is clearly not. Adding the keyboard dismiss on iOS itself is a workaround because `TextInput` is not focusable when `editable` is `false`.

https://github.com/Expensify/react-native-picker-select/blob/77cc9d42c474a693755941b10ee4c2d6f50e5346/src/index.js#L271-L273

<details>
<summary>Minimal code to reproduce</summary>

I can't provide a snack expo because `react-native-picker` requires native component.

```
import * as React from 'react';
import { Text, TextInput, View } from 'react-native';
import Picker from 'react-native-picker-select';

export default function App() {
  const [isFocused, setFocused] = React.useState(false);
  return (
    <View style={{padding: 16}}>
      <TextInput placeholder="test" style={{borderBottomWidth: 2, borderBottomColor: isFocused ? 'green' : 'black', marginBottom: 8}} onFocus={() => setFocused(true)} onBlur={() => setFocused(false)} />
      <Picker items={[{value: 1, label: 'A'}, {value: 2, label: 'B'}]} onValueChange={() => {}} />
    </View>
  );
}
```
</details>

### What changes do you think we should make in order to solve the problem?


You may think that we can do the same thing by dismissing the keyboard on picker click, but we can't. We need to do it on focus (focus here does not really a focus, I will explain later). As I said before, Android Picker component can be a `TextInput` or Android native Spinner and we use the former one because we set `useNativeAndroidPickerStyle` props to `false` and the input is wrapped by either `View` or `TouchableOpacity`, but in our case `View` is used.

https://github.com/Expensify/react-native-picker-select/blob/77cc9d42c474a693755941b10ee4c2d6f50e5346/src/index.js#L600-L604

https://github.com/Expensify/react-native-picker-select/blob/77cc9d42c474a693755941b10ee4c2d6f50e5346/src/index.js#L502-L541

The reason we can't dismiss the keyboard on press is because the `onPress` will never be called. I think this is related to these issues https://github.com/lawnstarter/react-native-picker-select/issues/354 https://github.com/Expensify/App/issues/13920. So, forget about the `onPress` and use `onFocus` props from react-native-picker Picker component instead. The `onFocus` prop is only for Android https://github.com/react-native-picker/picker#onblur.

I mentioned earlier that it is not really a focus, but just a click listener. Let see when the `onFocus` props is called.

https://github.com/react-native-picker/picker/blob/267de73e45b8fd8142316a1c43c94c9eec02ecb5/android/src/main/java/com/reactnativecommunity/picker/ReactPicker.java#L144-L152

From the link above, we can see the react picker override the `performClick` function and will fire the `onFocus` callback on every click.

We have 2 options to solve this inconsistency.
1. Dismiss the keyboard inside our Picker `onFocus` props here.
https://github.com/Expensify/App/blob/618ce68066deae8f9dc08768eb3b7885a1f3a725/src/components/Picker.js#L191

2. Add the workaround to the `react-native-picker-select` by dismissing the keyboard on every `onFocus` event to every `Picker` component use to render on Android (inside `renderAndroidHeadless` and `renderAndroidNativePickerStyle`.

The first option is very simple and requires very little changes. I'm not sure with the 2nd options because the repo looks like is not maintained for a long time. Last release is on Nov 17, 2020. Maybe we can do it on our fork.

**Comment by tgolen** [comment]
Wow, thanks for all that information. It sounds promising. What I think we should do is:

1. Implement the `onFocus` fix in [our fork](https://github.com/Expensify/react-native-picker-select)
2. Also create a PR in the source repo to get it changed there

If `2` ever gets merged, then we would update our fork with the latest source from upstream.

**Comment by tgolen** [comment]
🟢 👍 on the proposal from @bernhardoj

**Comment by bernhardoj** [comment]
I have opened the PR in our fork here https://github.com/Expensify/react-native-picker-select/pull/5.

I'm trying to open the PR in the upstream, but the commit history shows all previous commits done in our fork. Does anyone know how can I open the PR with my commit only?

![image](https://user-images.githubusercontent.com/50919443/224222136-694e591e-0f07-4708-970a-ccdb68687297.png)

**Comment by mdneyazahmad** [comment]
This looks like a duplicate of https://github.com/Expensify/App/issues/14783, reported from different perspective.

**Comment by bernhardoj** [comment]
Thanks for the suggestion!

Here is the PR.

Our fork: https://www.github.com/Expensify/react-native-picker-select/pull/5
Upstream: https://www.github.com/lawnstarter/react-native-picker-select/pull/501

**Comment by tgolen** [comment]
Thanks! I'm dropping this down to weekly now that it's in review.

**Comment by MelvinBot** [comment]
@tgolen @anmurali @thesahindia this issue is now 4 weeks old and preventing us from maintaining WAQ, can you:
- Decide whether any proposals currently meet our guidelines and can be approved as-is today
- If no proposals meet that standard, please take this issue internal and treat it as one of your highest priorities
- If you have any questions, don't hesitate to start a discussion in #expensify-open-source

Thanks!

**Comment by MelvinBot** [comment]
Current assignee @thesahindia is eligible for the Internal assigner, not assigning anyone new.

**Comment by tgolen** [comment]
The PR to our fork has been merged. @bernhardoj I think you just need to create a PR to update the version of our fork in package.json now.

**Comment by bernhardoj** [comment]
PR is ready for review.

**Comment by MelvinBot** [comment]
`Reviewing` label has been removed, please complete the "BugZero Checklist".

**Comment by MelvinBot** [comment]
The solution for this issue has been :rocket: deployed to production :rocket: in version 1.2.88-2 and is now subject to a 7-day regression period :calendar:. Here is the list of pull requests that resolve this issue:

- https://github.com/Expensify/App/pull/16045


If no regressions arise, payment will be issued on 2023-03-31. :confetti_ball:

After the hold period is over and BZ checklist items are completed, please complete any of the applicable payments for this issue, and check them off once done.
- [ ] External issue reporter
- [ ] Contributor that fixed the issue
- [ ] Contributor+ that helped on the issue and/or PR

As a reminder, here are the bonuses/penalties that should be applied for any External issue:
- Merged PR within 3 business days of assignment - 50% **bonus**
- Merged PR more than 9 business days after assignment - 50% **penalty**

**Comment by MelvinBot** [comment]
@tgolen, @anmurali, @thesahindia Whoops! This issue is 2 days overdue. Let's get this updated quick!

**Comment by tgolen** [comment]
@anmurali Looks like this is just waiting on you.

**Comment by tgolen** [comment]
bump @anmurali

**Comment by anmurali** [comment]
Creating regression test - https://github.com/Expensify/Expensify/issues/275047

**Comment by tgolen** [comment]
I think this is completed now so closing it out.
